src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InitializeStaticFieldsInline.cs

```diff
@@ -45,39 +46,46 @@ public abstract class InitializeStaticFieldsInlineAnalyzer<TLanguageKindEnum> :
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(CA1810Rule, CA2207Rule);
 
-        protected abstract bool InitialiesStaticField(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancellationToken);
-        protected abstract TLanguageKindEnum AssignmentNodeKind { get; }
-
         public override void Initialize(AnalysisContext analysisContext)
         {
             analysisContext.EnableConcurrentExecution();
             analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
 
-            analysisContext.RegisterCodeBlockStartAction<TLanguageKindEnum>(codeBlockStartContext =>
+            analysisContext.RegisterOperationBlockStartAction(context =>
             {
-                if (!(codeBlockStartContext.OwningSymbol is IMethodSymbol methodSym) ||
-                    !methodSym.IsStatic ||
-                    methodSym.MethodKind != MethodKind.SharedConstructor)
+                if (!(context.OwningSymbol is IMethodSymbol method) ||
+                    !method.IsStatic ||
+                    method.MethodKind != MethodKind.StaticConstructor)
                 {
                     return;
                 }
 
                 var initializesStaticField = false;
-                codeBlockStartContext.RegisterSyntaxNodeAction(syntaxContext =>
+                context.RegisterOperationAction(context =>
                 {
-                    if (!initializesStaticField)
+                    if (initializesStaticField)
+                    {
+                        return;
+                    }
+
+                    var assignment = (IAssignmentOperation)context.Operation;
+
+                    if (assignment.Target is IFieldReferenceOperation fieldReference &&
+                        fieldReference.Member.IsStatic &&
+                        assignment.GetAncestor<IAnonymousFunctionOperation>(OperationKind.AnonymousFunction) is null)
                     {
-                        initializesStaticField = InitialiesStaticField(syntaxContext.Node, syntaxContext.SemanticModel, syntaxContext.CancellationToken);
+                        initializesStaticField = true;
                     }
-                }, AssignmentNodeKind);
+                }, OperationKind.SimpleAssignment);
 
-                codeBlockStartContext.RegisterCodeBlockEndAction(codeBlockEndContext =>
+                context.RegisterOperationBlockEndAction(context =>
                 {
                     if (initializesStaticField)
                     {
-                        DiagnosticDescriptor descriptor = methodSym.ContainingType.IsReferenceType ? CA1810Rule : CA2207Rule;
-                        Diagnostic diagnostic = Diagnostic.Create(descriptor, methodSym.Locations[0], methodSym.ContainingType.Name);
-                        codeBlockEndContext.ReportDiagnostic(diagnostic);
+                        context.ReportDiagnostic(
```

Comment:
I think we will end up reporting a diagnostic if a type has one static field that is assigned a delegate and one static field with a simple initializer. The correct fix would be to use a separate bool variable to track if any field is assigned a delegate, and if so bail out from the block end action.

Reply:
Theoretically, we can still flag methods which have at least one static field initialized directly, but if user already needs a static cctor to initialize one of the fields inside a lambda, they might prefer initializing all static fields in the static cctor. Best to just bail out for the case where any one field is assigned inside a lambda.